docarray/utils/apply.py

```diff
@@ -0,0 +1,296 @@
+        da[indices] = batch
+
+
+def _map_batch(
+    da: T,
+    func: Callable[[T], T],
+    batch_size: int,
+    backend: str = 'thread',
+    num_worker: Optional[int] = None,
+    shuffle: bool = False,
+    pool: Optional[Union[Pool, ThreadPool]] = None,
+    show_progress: bool = False,
+) -> Generator[T, None, None]:
+    """
+    Return an iterator that applies `func` to every **minibatch** of iterable in parallel,
+    yielding the results.
+    Each element in the returned iterator is an :class:`AnyDocumentArray`.
+
+    .. seealso::
+        - To return :class:`DocumentArray`, please use :func:`apply_batch`.
+
+    :param batch_size: Size of each generated batch (except the last one, which might
+        be smaller).
+    :param shuffle: If set, shuffle the Documents before dividing into minibatches.
+    :param func: a function that takes an :class:`AnyDocumentArray` as input and outputs
+        an :class:`AnyDocumentArray` of the same Document type as the input.
+    :param backend: `thread` for multithreading and `process` for multiprocessing.
+        Defaults to `thread`.
+        In general, if `func` is IO-bound then `thread` is a good choice.
+        On the other hand, if `func` is CPU-bound, then you may use `process`.
+        In practice, you should try yourselves to figure out the best value.
+        However, if you wish to modify the elements in-place, regardless of IO/CPU-bound,
+        you should always use `thread` backend.
+        Note that computation that is offloaded to non-python code (e.g. through np/torch/tf)
+        falls under the "IO-bound" category.
+
+        .. warning::
+            When using `process` backend, your `func` should not modify elements in-place.
+            This is because the multiprocessing backend passes the variable via pickle
+            and works in another process.
+            The passed object and the original object do **not** share the same memory.
+
+    :param num_worker: the number of parallel workers. If not given, then the number of CPUs
+        in the system will be used.
+    :param show_progress: show a progress bar
+    :param pool: use an existing/external pool. If given, `backend` is ignored and you will
+        be responsible for closing the pool.
+
+    :yield: DocumentArrays returned from `func`
+    """
+    if backend == 'process' and _is_lambda_or_partial_or_local_function(func):
+        raise ValueError(
+            f'Multiprocessing does not allow functions that are local, lambda or partial: {func}'
+        )
+
+    context_pool: Union[nullcontext, Union[Pool, ThreadPool]]
+    if pool:
+        p = pool
+        context_pool = nullcontext()
+    else:
+        p = _get_pool(backend, num_worker)
+        context_pool = p
+
+    with context_pool:
+        imap = p.imap(func, da._batch(batch_size=batch_size, shuffle=shuffle))
+        for x in track(
+            imap, total=ceil(len(da) / batch_size), disable=not show_progress
```

Comment:
what happened if the batch size is not a multiple of the len of da ? Can u add a test to cover this ?